pkg/component/controller/cplb_unix.go

```diff
@@ -0,0 +1,328 @@
+		return fmt.Errorf("failed to configure dummy interface: %v", err)
+	}
+
+	if err := k.Config.ValidateVRRPInstances(); err != nil {
+		return fmt.Errorf("failed to validate VRRP instances: %v", err)
+	}
+	if err := k.generateKeepalivedTemplate(); err != nil {
+		return fmt.Errorf("failed to generate keepalived template: %v", err)
+
+	}
+
+	k.log.Infoln("Starting keepalived")
+	k.supervisor = &supervisor.Supervisor{
+		Name:    "keepalived",
+		BinPath: assets.BinPath("keepalived", k.K0sVars.BinDir),
+		Args: []string{
+			"--dont-fork",
+			"--use-file",
+			k.K0sVars.KeepalivedConfigFile,
+		},
+		RunDir:  filepath.Dir(k.K0sVars.KeepalivedConfigFile),
+		DataDir: filepath.Dir(k.K0sVars.KeepalivedConfigFile),
+		UID:     k.uid,
+	}
+	return k.supervisor.Supervise()
+}
+
+// Stops keepalived and cleans up the virtual IPs. This is done so that if the
+// k0s controller is stopped, it can still reach the other APIservers on the VIP
+func (k *Keepalived) Stop() error {
+	k.log.Infof("Stopping keepalived")
+	if err := k.supervisor.Stop(); err != nil {
+		// Failed to stop keepalived. Don't delete the VIP, just in case.
+		return fmt.Errorf("failed to stop keepalived: %v", err)
+	}
+
+	k.log.Infof("Deleting dummy interface")
+	link, err := netlink.LinkByName(dummyLinkName)
+	if err != nil {
+		if errors.As(err, &netlink.LinkNotFoundError{}) {
+			return nil
+		}
+		k.log.Errorf("failed to get link by name %s. Attempting to delete it anyway: %v", dummyLinkName, err)
+		link = &netlink.Dummy{
+			LinkAttrs: netlink.LinkAttrs{
+				Name: dummyLinkName,
+			},
+		}
+	}
+	return netlink.LinkDel(link)
+}
+
+// configureDummy creates the dummy interface and sets the virtual IPs on it.
+func (k *Keepalived) configureDummy() error {
+	vips := []string{}
+	for _, vi := range k.Config.VRRPInstances {
+		for _, vip := range vi.VirtualIPs {
+			vips = append(vips, vip)
+		}
+	}
+
+	if len(vips) > 0 {
+		k.log.Infof("Creating dummy interface")
+		if err := k.ensureDummyInterface(dummyLinkName); err != nil {
+			k.log.Errorf("failed to create dummy interface: %v", err)
+		}
+		// If the dummy interface fails, attempt to define the addresses just
+		// in case.
+		if err := k.ensureLinkAddresses(dummyLinkName, vips); err != nil {
+			return fmt.Errorf("failed to ensure link addresses: %v", err)
+		}
+	}
+	return nil
+}
+
+func (k *Keepalived) ensureDummyInterface(linkName string) error {
+	link, err := netlink.LinkByName(linkName)
+	if err != nil {
+		// There are multiple reasons why the link may not be returned besides
+		// it not existing. If we don't know what failed log it and attempt to
+		// create the link anyway.
+		if !errors.As(err, &netlink.LinkNotFoundError{}) {
+			k.log.Warnf("failed to get link by name %s. Attempting to create it anyway: %v", linkName, err)
+		}
+		return k.createDummyInterface(linkName)
+	}
+
+	if _, isDummy := link.(*netlink.Dummy); isDummy {
+		return nil
+	}
+
+	if err = netlink.LinkDel(link); err != nil {
```

Comment:
Is delete done here too in case of user changes the address and we need to re-create the dummy interface?

Reply:
This would happen if the interface exists but it's not of type dummy, I don't think it can actually happen but since it's there I'd say it doesn't do any harm. I added a comment explaining why it's there.